master/internal/api_experiment.go

```diff
@@ -708,3 +710,172 @@ func (a *apiServer) TrialsSnapshot(req *apiv1.TrialsSnapshotRequest,
+	case ByTrainingLength:
+		return a.m.db.TopTrialsByTrainingLength(experimentID, maxTrials, s.Metric, s.SmallerIsBetter)
+	default:
+		panic("Invalid state in trial sampling")
+	}
+}
+
+func (a *apiServer) TrialsSample(req *apiv1.TrialsSampleRequest,
+	resp apiv1.Determined_TrialsSampleServer) error {
+	experimentID := int(req.ExperimentId)
+	maxTrials := int(req.MaxTrials)
+	if maxTrials == 0 {
+		maxTrials = 25
+	}
+	maxDatapoints := int(req.MaxDatapoints)
+	if maxDatapoints == 0 {
+		maxDatapoints = 1000
+	}
+	startBatches := int(req.StartBatches)
+	endBatches := int(req.EndBatches)
+	if endBatches <= 0 {
+		endBatches = math.MaxInt32
+	}
+
+	metricName := req.MetricName
+	metricType := req.MetricType
+	if metricType == apiv1.MetricType_METRIC_TYPE_UNSPECIFIED {
+		return errors.New("must specify a metric type")
+	}
+	if metricName == "" {
+		return errors.New("must provide metric name")
+	}
+
+	config, err := a.m.db.ExperimentConfig(experimentID)
+	if err != nil {
+		return errors.Wrapf(err, "error fetching experiment config from database")
+	}
+	searcherConfig := config.Searcher
+
+	trialCursors := make(map[int32]time.Time)
+	currentTrials := make(map[int32]bool)
+	var zeroTime time.Time
+	for {
+		var response apiv1.TrialsSampleResponse
+		var promotedTrials []int32
+		var demotedTrials []int32
+		var trials []*apiv1.TrialsSampleResponse_Trial
+
+		seenThisRound := make(map[int32]bool)
+
+		var metricSeries []lttb.Point
+		var endTime time.Time
+		trialIDs, err := a.topTrials(experimentID, maxTrials, searcherConfig)
+		if err != nil {
+			return errors.Wrapf(err, "error determining top trials")
+		}
+		for _, trialID := range trialIDs {
+			var trial apiv1.TrialsSampleResponse_Trial
+			trial.TrialId = trialID
+
+			if _, current := currentTrials[trialID]; !current {
+				promotedTrials = append(promotedTrials, trialID)
+				currentTrials[trialID] = true
+
+				var trialConfig *model.Trial
+				trialConfig, err = a.m.db.TrialByID(int(trialID))
+				if err != nil {
+					return errors.Wrapf(err, "error fetching trial metadata")
+				}
+				trial.Hparams = protoutils.ToStruct(trialConfig.HParams)
+			}
+			seenThisRound[trialID] = true
+
+			startTime, seenBefore := trialCursors[trialID]
+			if !seenBefore {
+				startTime = zeroTime
+			}
+			if metricType == apiv1.MetricType_METRIC_TYPE_TRAINING {
+				metricSeries, endTime, err = a.m.db.TrainingMetricsSeries(trialID, startTime,
+					metricName, startBatches, endBatches)
+			} else {
+				metricSeries, endTime, err = a.m.db.ValidationMetricsSeries(trialID, startTime,
+					metricName, startBatches, endBatches)
+			}
+			if err != nil {
+				return errors.Wrapf(err, "error fetching time series of metrics")
+			}
+			if len(metricSeries) > 0 {
+				// if we get empty results, the endTime is incorrectly zero
+				trialCursors[trialID] = endTime
+			}
+			if !seenBefore {
+				metricSeries = lttb.Downsample(metricSeries, maxDatapoints)
```

Comment:
question: curious, why do we only downsample if it's the first time seeing it? is the idea we dont want to send way too many metrics at once if a long running trial suddenly pops into top trials?

Reply:
Exactly. It would be quite common for someone to load a page with an experiment that has run for days and get millions of data points. Very resource intensive to render all of those data points at once. It would be less common for someone to leave the page open constantly rendering for long enough to accumulate a comparable amount of data (and even then, it would be spread out over a longer period of time, and down-sampling the stream while still providing frequent updates would be quite a bit messier).

Reply:
cool. just making sure I understood the intent.